File: fork_test.py
from os import fork
from time import sleep


SHARED_STATE = 0


def set_shared_state(st):
    global SHARED_STATE
    SHARED_STATE = st


def main():
    print('before fork')
    pid = fork()
    print('after fork pid=%d' % pid)
    if pid != 0:
        print('setting shared state to %d' % pid)
        set_shared_state(pid)
    else:
        sleep(1)
    global SHARED_STATE
    print('pid %d says shared state is %d' % (pid, SHARED_STATE))


if __name__ == '__main__':
    main()

File: fork_test.cc
#include <unistd.h>
#include <iostream>

pid_t shared_state = 0;

int main(int argc, char** argv, char** envp) {
    std::cout << "before fork" << std::endl;
    pid_t pid = fork();
    std::cout << "after fork pid " << pid << std::endl;
    if (pid != 0) {
        shared_state = pid;
    } else {
        usleep(1000000);
    }
    std::cout << "pid " << pid << " says shared state is " << shared_state << std::endl;
    return 0;
}
